### PR TITLE
Sign in

### PR DESCRIPTION
**Pull request by fmontoy**
## Summary

Creating the sign-in endpoint with unit tests.

## Known Issues

No known issues

## Trello Card

[Trello Card](https://trello.com/c/YorzQ1IZ/3-sign-in)

### CODE REVIEW

**Comment by raparicio6** [comment]
Builds failed

File: app/constants.js [hunk 1]
```diff
@@ -1 +1,2 @@
 exports.SALTROUNDS = 10;
+exports.SECRET = 'wolox';
```
Comment (raparicio6) [inline app/constants.js]: secrets always are env vars

File: app/controllers/users.js [hunk 2]
```diff
@@ -1,7 +1,17 @@
...
 exports.createUser = (req, res, next) =>
   createUser(req.body)
     .then(user => res.send(lodash.pick(user, ['name', 'last_name', 'email'])))
     .catch(next);
+
+exports.signIn = (req, res, next) =>
+  signInUser(req.body)
+    .then(() => {
+      const token = jwt.encode({ username: req.body.email }, SECRET);
+      res.send({ token });
```
Comment (raparicio6) [inline app/controllers/users.js]: we always return the last line

File: app/schemas/user.js [hunk 3]
```diff
@@ -43,3 +43,28 @@ exports.userSchema = {
...
+  email: {
+    in: ['body'],
+    matches: {
+      options: /^[a-zA-Z0-9]+@wolox.([A-Za-z])+$/,
+      errorMessage: errors.badRequestError('The email must have the wolox domain')
+    },
+
+    isLength: {
+      options: { min: 9 },
+      errorMessage: errors.badRequestError('The email cannot be empty')
```
Comment (raparicio6) [inline app/schemas/user.js]: cannot be empty or cannot be less than 9 characters?

File: app/services/users.js [hunk 4]
```diff
@@ -2,9 +2,39 @@ const { User } = require('../models');
 const errors = require('../errors');
 const bcrypt = require('bcryptjs');
 const { SALTROUNDS } = require('../constants');
-
+const logger = require('../logger');
```
Comment (raparicio6) [inline app/services/users.js]: we put a breakline after requires

File: app/services/users.js [hunk 5]
```diff
@@ -2,9 +2,39 @@ const { User } = require('../models');
...
 exports.createUser = user =>
   bcrypt.hash(user.password, SALTROUNDS).then(hash => {
     user.password = hash;
     return User.create(user).catch(error => Promise.reject(errors.databaseError(error.message)));
   });
+
+exports.signInUser = ({ email, password }) => {
+  logger.info('Trying to sign-in the user');
```
Comment (raparicio6) [inline app/services/users.js]: here you should log which user you are logging

File: app/services/users.js [hunk 6]
```diff
@@ -2,9 +2,39 @@ const { User } = require('../models');
...
 const { SALTROUNDS } = require('../constants');
-
+const logger = require('../logger');
 exports.createUser = user =>
   bcrypt.hash(user.password, SALTROUNDS).then(hash => {
     user.password = hash;
     return User.create(user).catch(error => Promise.reject(errors.databaseError(error.message)));
   });
+
+exports.signInUser = ({ email, password }) => {
```
Comment (raparicio6) [inline app/services/users.js]: this method should be here. we dont put bussiness logic in services. 
here you should only put a method to do the findOne in the db

File: app/controllers/users.js [hunk 7]
```diff
@@ -1,7 +1,40 @@
...
+
+exports.signIn = (req, res, next) =>
+  signInUser(req.body)
+    .then(user => {
+      if (user) {
+        return bcrypt
+          .compare(req.body.password, user.password)
+          .then(passwordExists => {
+            if (!passwordExists) {
+              next(errors.badRequestError('Wrong password!'));
```
Comment (raparicio6) [inline app/controllers/users.js]: what are you missing here? :thinking:

File: app/controllers/users.js [hunk 8]
```diff
@@ -1,7 +1,40 @@
...
+            }
+            logger.info('User exists and passwords are matching');
+            const token = jwt.encode({ username: req.body.email }, secret);
```
Comment (raparicio6) [inline app/controllers/users.js]: encode is not a promise?

File: app/services/users.js [hunk 9]
```diff
@@ -2,9 +2,19 @@ const { User } = require('../models');
...
 const { SALTROUNDS } = require('../constants');
+const logger = require('../logger');
 
 exports.createUser = user =>
   bcrypt.hash(user.password, SALTROUNDS).then(hash => {
     user.password = hash;
     return User.create(user).catch(error => Promise.reject(errors.databaseError(error.message)));
   });
+
+exports.signInUser = ({ email }) => {
```
Comment (raparicio6) [inline app/services/users.js]: mmm why do you call this signInUser? what do you think about findUser?

File: test/user.spec.js [hunk 10]
```diff
@@ -52,3 +58,42 @@ describe('User Creation', () => {
...
+    request(app)
+      .post('/users')
+      .send(user)
+      .then(() =>
+        request(app)
+          .post('/users/sessions')
+          .send(signIn)
+          .then(res => {
+            expect(res.statusCode).toEqual(200);
+            expect(jwt.decode(res.body.token, secret).username).toEqual(signIn.email);
```
Comment (raparicio6) [inline test/user.spec.js]: we prefer using strict equal

File: test/utils.js [hunk 11]
```diff
@@ -4,3 +4,8 @@ exports.user = {
   email: 'joedoe@wolox.co',
   password: 'holamundo2019'
 };
+
+exports.signIn = {
```
Comment (raparicio6) [inline test/utils.js]: what do you think about signInData?

File: app/controllers/users.js (hunk 8 above)
Comment (fmontoy) [reply app/controllers/users.js]: Based in the documentation, is not a promise https://www.npmjs.com/package/jwt-simple

File: test/user.spec.js (hunk 10 above)
Comment (fmontoy) [reply test/user.spec.js]: I used strictEqual and toBe, it seems that toBe works better for testing primitive types :) https://stackoverflow.com/questions/45195025/what-is-the-difference-between-tobe-and-toequal-in-jest

File: app/controllers/users.js [hunk 12]
```diff
@@ -1,7 +1,40 @@
...
 exports.createUser = (req, res, next) =>
   createUser(req.body)
     .then(user => res.send(lodash.pick(user, ['name', 'last_name', 'email'])))
     .catch(next);
+
+exports.signIn = (req, res, next) =>
+  findUser(req.body)
+    .then(user => {
+      if (user) {
+        return bcrypt
```
Comment (raparicio6) [inline app/controllers/users.js]: I would put this in other file to have this code generic

File: app/services/users.js [hunk 13]
```diff
@@ -2,9 +2,19 @@ const { User } = require('../models');
...
 const { SALTROUNDS } = require('../constants');
+const logger = require('../logger');
 
 exports.createUser = user =>
   bcrypt.hash(user.password, SALTROUNDS).then(hash => {
     user.password = hash;
     return User.create(user).catch(error => Promise.reject(errors.databaseError(error.message)));
   });
+
+exports.findUser = ({ email }) => {
```
Comment (raparicio6) [inline app/services/users.js]: this should be generic to filter with any property

File: app/services/users.js [hunk 14]
```diff
@@ -2,9 +2,19 @@ const { User } = require('../models');
...
+  logger.info(`Trying to sign-in the user with email: ${email}`);
```
Comment (raparicio6) [inline app/services/users.js]: this log should be here?

File: app/services/users.js [hunk 15]
```diff
@@ -8,3 +8,15 @@ exports.createUser = user =>
...
+exports.findUser = params =>
+  User.findOne({
+    where: {
+      ...params
+    }
+  }).then(user => {
+    if (user) {
+      return Promise.resolve(user);
+    }
+    return Promise.reject(errors.badRequestError('No user with that email'));
```
Comment (raparicio6) [inline app/services/users.js]: you need to put this in the service?

Comment (raparicio6) [reply app/services/users.js]: what do you think it would be most properly?

File: app/services/users.js [hunk 16]
```diff
@@ -8,3 +8,15 @@ exports.createUser = user =>
     user.password = hash;
     return User.create(user).catch(error => Promise.reject(errors.databaseError(error.message)));
   });
+
+exports.findUser = params =>
+  User.findOne({
+    where: {
```
Comment (raparicio6) [inline app/services/users.js]: { ...object } = object

**Pull request merged**